optimade/server/entry_collections.py

```diff
@@ -119,19 +119,18 @@ def find(
             fields = set(params.response_fields.split(","))
         else:
             fields = all_fields.copy()
-        results = []
 
-        # build up set of included data with unique IDs
-        included = []
-        id_set = set()
+        results = []
+        included = {}
         for doc in self.collection.find(**criteria):
             results.append(self.resource_cls(**self.resource_mapper.map_back(doc)))
-            for reference in (
-                doc.get("relationships", {}).get("references", {}).get("data", [])
-            ):
-                if reference["id"] not in id_set:
-                    included.append(reference)
-                    id_set.add(reference["id"])
+            # collapse list of references into dict by ID and take only one per ID
+            refs = doc.get("relationships", {}).get("references", {}).get("data", [])
+            for ref in refs:
+                # could check here and raise a warning if any IDs clash
```

Comment:
It's not needed. This way we will get a minimum set as well. Even though we may set the `included` key for a specific `"id"` several times.